### PR TITLE
Add Docker for Mac host time sync daemon

### PR DESCRIPTION
**Pull request by djs55**
**- What I did**

Cause the Docker for Mac VM clock to be resynchronised when the host awakes from sleep.

**- How I did it**

Host-side code on Docker for Mac `connect`s to an `AF_VSOCK` socket (port `0xf3a4`) whenever it detects that the Mac has resumed from sleep. This connection is received by a daemon called `host-timesync-daemon` in the VM which reads `/dev/rtc0` and calls `settimeofday` (this requires `CAP_SYS_TIME`)

The first version of this code was originally written in C by @MagnusS -- I've ported it to Go in this PR.

**- How to verify it**

- `moby build docker-for-mac.yml`
- `linuxkit run docker-for-mac`
- `tail -f /var/log/host-timesync-daemon.log `
- on the Host in the `blueprints` directory: `echo "00000003.0000f3a4" | nc -U ./docker-for-mac-state/connect `
- observe in the log entries like
```
2017/06/28 16:48:58 Connection to: ffffffff:f3a4 from: 00000002.04000003
2017/06/28 16:48:58 Setting system clock to 2017-06-28 16:48:59 +0000 UTC
2017/06/28 16:49:01 Connection to: ffffffff:f3a4 from: 00000002.04000004
2017/06/28 16:49:01 Setting system clock to 2017-06-28 16:49:01 +0000 UTC
```

### CODE REVIEW

**Comment by justincormack** [comment]
I don't think you need to shell out, its just `ioctl` `RTC_RD_TIME` on `/dev/rtc` (see man 4 rtc) and then `settimeofday`, should be able to do all that in Go...

**Comment by djs55** [comment]
@justincormack I've removed `hwclock` and used the `ioctl` as you suggested. Let me know what you think!

File: pkg/host-timesync-daemon/main.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		// that case we assume the platform is capable of providing a PV clock and
+		// we don't use this code anyway.
+		d := time.Date(int(t.tmYear+1900), time.Month(t.tmMon+1), int(t.tmMday), int(t.tmHour), int(t.tmMin), int(t.tmSec), 0, time.UTC)
+		log.Printf("Setting system clock to %s", d)
+		tv := syscall.Timeval{
+			Sec:  int64(d.Unix()),
+			Usec: 0, // the RTC only has second granularity
+		}
+		if err = syscall.Settimeofday(&tv); err != nil {
+			log.Printf("Unexpected failure from Settimeofday: %#v", err)
```
Comment (justincormack) [inline pkg/host-timesync-daemon/main.go]: you only need `%v` here

File: pkg/host-timesync-daemon/main.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/linuxkit/virtsock/pkg/vsock"
```
Comment (justincormack) [inline pkg/host-timesync-daemon/main.go]: this should be at the end, see other Go files. (doesn't `go fmt` fix this?)

Comment (djs55) [inline pkg/host-timesync-daemon/main.go]: I tried moving it to the end, then `gofmt -w -s` puts it back! I was confused by this so I did a little more experimentation. It seems that inserting a blank line between the block of regular imports and the github.com ones has the desired effect. I think it must be alphabetical within each newline-delimited block. I'll modify the patch.

File: pkg/host-timesync-daemon/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+### host-timesync-daemon
+
+Some hypervisors (e.g. hyperkit / xhyve) don't provide a good way to keep
+the VM's clock in sync with the Host's clock. NTP will usually keep the
+clocks together, but after a the host or VM is suspended and resumed the
+clocks can be suddenly too far apart for NTP to work properly.
+
+This simple daemon listens on an AF_VSOCK port and resynchronises the VM
+clock from the virtualised hardware clock via querying /dev/rtc0 and calling
+`settimeofday`.
```
Comment (justincormack) [inline pkg/host-timesync-daemon/README.md]: you should say what has to be on the other end of the socket...

Comment (justincormack) [inline pkg/host-timesync-daemon/README.md]: Its not obvious its just "connect to it"

File: pkg/host-timesync-daemon/main.go (hunk 2 above)
Comment (ijc) [inline pkg/host-timesync-daemon/main.go]: Yes, it's alpha in each "paragraph". Most projects seem to separate stdlib ones like `fmt` etc from third party ones with a blank line. Some block them all together, others group them by functional area etc.

File: pkg/host-timesync-daemon/main.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		log.Fatalf("RTC_RD_TIME failed: %v", errno)
+	}
+	return result
+}
+
+func main() {
+	// host-timesync-daemon -cid <cid> -port <port>
+
+	cid := flag.Int("cid", 0, "AF_HVSOCK CID to listen on")
+	port := flag.Int("port", 0, "AF_HVSOCK port to listen on")
```
Comment (justincormack) [inline pkg/host-timesync-daemon/main.go]: I think both those lines should say `AF_VSOCK` not `HVSOCK` (especially as hvsock is going away)

File: pkg/host-timesync-daemon/main.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "%s: set the time after an AH_VSOCK connection is received.\n\n", os.Args[0])
```
Comment (justincormack) [inline pkg/host-timesync-daemon/main.go]: typo `AH_VSOCK`

**Comment by MagnusS** [comment]
This needs a rebase after #2137. The blueprint is now in `blueprints/docker-for-mac/base.yml`

**Comment by djs55** [comment]
I've rebased to current master. I think this is ready for merging although we'll need to create the image repo, build sign and push a tag, and then update the .yml to refer to it.

File: pkg/host-timesync-daemon/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,22 @@
+FROM linuxkit/alpine:630ee558e4869672fae230c78364e367b8ea67a9 AS mirror
```
Comment (rn) [inline pkg/host-timesync-daemon/Dockerfile]: should be `linuxkit/alpine:2e362f4459ba4491655061cccdd2fcc7a2de5eb3` now that #2154 is merged

**Comment by MagnusS** [comment]
Anything left on this except the hub repo access? Would be good to get this in as it's one of the last Docker for Mac components.

**Comment by djs55** [comment]
I've corrected the hash spotted by @rn so I now think this is ready for merging again (although I could be wrong!)

It still needs someone to create a `linuxkit/host-timesync-daemon` repo, build, sign, push and then update the hash in the blueprint -- I can't do that myself yet unfortunately.

**Comment by riyazdf** [comment]
@djs55 I made the repo/set up signing, but am running into a Makefile issue. It doesn't pick up the correct image name because the include statement needs to come after the image variable definition:

```
host-timesync-daemon 🐳 $ make
docker pull linuxkit/:d8a65b428a64cbdbe7560910b16f3ccb6fd35b06 || docker build --network=none -t linuxkit/:d8a65b428a64cbdbe7560910b16f3ccb6fd35b06 .
invalid reference format
invalid argument "linuxkit/:d8a65b428a64cbdbe7560910b16f3ccb6fd35b06" for t: invalid reference format
See 'docker build --help'.
make: *** [tag] Error 125
```

Swapping things around and adding `NETWORK=1` so it can clone with this diff makes things work:
```
diff --git a/pkg/host-timesync-daemon/Makefile b/pkg/host-timesync-daemon/Makefile
index d09f8f7e..28d66a17 100644
--- a/pkg/host-timesync-daemon/Makefile
+++ b/pkg/host-timesync-daemon/Makefile
@@ -1,4 +1,5 @@
-include ../package.mk
-
 IMAGE=host-timesync-daemon
 DEPS=$(wildcard *.go)
+NETWORK=1
+
+include ../package.mk
```

**Comment by djs55** [comment]
@riyazdf sorry about that-- I've forced pushed a new version with your diff applied. Thanks!

**Comment by riyazdf** [comment]
no problem @djs55!  Signed and pushed `linuxkit/host-timesync-daemon:e6c15e6ef75d302c1c22827bac249598fb365a83`

**Pull request merged**